File: 3380-shortest-subarray-with-or-at-least-k-ii/shortest-subarray-with-or-at-least-k-ii.py
class Solution:
    def minimumSubarrayLength(self, nums: List[int], k: int) -> int:
        result = float("inf")
        bits = [0] * 32

        def bitOperation(bits, num, diff):
            for i in range(32):
                if num & (1 << i):
                    bits[i] += diff

        def getInteger(bits):
            temp = 0
            for i in range(32):
                if bits[i] > 0:
                    temp += (1 << i)
            return temp

        l = 0
        

        for r in range(len(nums)):
            bitOperation(bits, nums[r], 1)
            curr = getInteger(bits)
            while curr >= k and l <= r:
                result = min(result, r - l + 1)
                bitOperation(bits, nums[l], -1)
                curr = getInteger(bits)
                l += 1
        return -1 if result == float("inf") else result

File: 3380-shortest-subarray-with-or-at-least-k-ii/shortest-subarray-with-or-at-least-k-ii.cpp
class Solution {
public:
    void updateBits(int* bits, int nums, int diff){
        for (int i = 0; i < 32; i++)
            if ((nums & (1 << i)) > 0)
                bits[i] += diff;
    }
    int getInteger(int* bits){
        int temp = 0;
        for (int i = 0; i < 32; i++)
            if (bits[i] > 0)
                temp+= (1 << i);
        return temp;
    }

    int minimumSubarrayLength(vector<int>& nums, int k) {
        int bits[32] = {0};
        int l = 0;
        int result = INT_MAX;
        for (int r = 0; r < nums.size(); r++){
            updateBits(bits, nums[r], 1);
            int curr = getInteger(bits);
            while (curr >= k && l <= r){
                result = min(result, r - l + 1);
                updateBits(bits, nums[l], -1);
                curr = getInteger(bits);
                l += 1;
            }
        }
        return result == INT_MAX ? -1 : result;
    }
};